### DISCUSSION TITLE
How to deal with separate frontend and backend

### DISCUSSION THREAD

**Discussion by HughWarrington** [joyfullservice/msaccess-vcs-addin#540] [Q&A] [answered]
I'm working on a project with separate frontend and backend. The frontend has linked tables which point to the backend. At runtime, the frontend is designed so that the user can select one of many backends, and uses `CurrentDb().Connect` and `.RefreshLink` to link the tables at startup before any queries are made.

When I use *Build From Source* in the plugin, I get this error for each linked table:

```
-------------------------------------
ERROR: Could not link table 'Foo'
Source: clsDbTableDef.ImportLinkedTable
-------------------------------------
Linked table object not found in C:\Users\...
Connection String: ;DATABASE=C:\Users\...
```

What's the recommended approach here? The tables should not to be linked to anything until runtime.

**Comment by joyfullservice** [comment, marked as answer]
Is the back end available during the development of the front end? It appears that the linked table target could not be found during the build. A linked table needs to be linked to "something" during creation, or the link will fail. (As you noted above.)

There are a few approaches you could take with this. One approach would be to include the front and back end tables in the same repository, and build the back end database before building the front end database. This works well when the front and back end have more of a tight coupling where changes need to be made together.

Another approach would be to tie into the `BeforeExport` hook and relink the back end tables to a known set that will always be available during build. These could be empty tables in a database on a network share (or SQL server) but will allow the database to be successfully built. This way any build from source would always use the set of known tables, regardless of whether the production tables were available to the build environment.

While a linked table can still _exist_ in a database even if the target cannot be accessed, I don't think you can create a linked table without actually having a valid target. I suppose you could even create a temporary target for the build, tying into the `BeforeBuild` hook and delete it afterwards, but that seems like a lot of extra work.  😄

**Reply by HughWarrington** [reply]
Got it, thanks for the info.

My frontend and backend are in the same repo. So I think I can ensure during development that the frontend is always pointing to the backend in the same repo using a relative path, even though this would never be the case at runtime.

**Reply by HughWarrington** [reply]
My repo currently looks like this:

```
repo/
    backend.src/
    frontend.src/
    backend.mdb
    frontend.mdb
```

I edited my linked tbldefs in the frontend to look like this (e.g. file `frontend\tbldefs\AnsDate.json`):

```
﻿{
  "Info": {
    "Class": "clsDbTableDef",
    "Description": "Linked Table"
  },
  "Items": {
    "Name": "AnsDate",
    "Connect": ";DATABASE=backend.mdb",
    "SourceTableName": "AnsDate",
    "Attributes": 1073741824
  }
}
```

Building the frontend now works, and the linked tables are linked without error since backend.mdb exists at build time. However there is a new problem. Now when I export the _frontend_, it contains a new .sql file for each _backend_ table. So in the example, a new file `frontend\tbldefs\AnsDate.sql` is created. Also `frontend\tbldefs\AnsDate.json` contains an additional attribute `PrimaryKey`.

Now all those tbldefs exist in the repo twice: once under backend.src and again under frontend.src. Do you have a suggestion about how to deal with this?

**Reply by joyfullservice** [reply]
Hmm... Perhaps we shouldn't be generating `.sql` files for _linked_ tables. 🤔 In one sense, it is handy to immediately know if something changes in the linked back-end table, but this is a bit redundant when the back-end table is being managed through another export. Since the structure of a linked table can't be changed in the front-end database anyway, I am leaning towards filtering the `.sql` export to not run on linked tables.

**Reply by HughWarrington** [reply]
That sounds sensible to me. If there is anyone out there relying on the current behaviour, you could always make exporting linked tables optional behaviour.

**Reply by hecon5** [reply]
So, this will have pretty significant (and ... unexpected) consequences for people not realizing that linked tables won't export.

I actually think this is the right way, but this seems actually like a breaking change, being that the exports will no longer be provided.

While you do have a tool to export the SQL, IMHO there's going to be folks out there who do not realize this change means their linked table (and views) won't be exported any more. We've actually noticed backend changes that were not reviewed this way before...

IMHO: if we do stop exporting the linked tables, we need to point users to the backend export tooling as a recourse, or they're going to see all the deleted export files and freak.

**Reply by joyfullservice** [reply]
@hecon5 - Just to clarify, nothing is changing with the export of linked tables. You can still export and build your project with linked tables. They are still just as supported by the add-in. What is changing is that the optional one-way helper export of a SQL representation of tables will no longer apply to linked tables. The SQL view is purely for visual reference purposes, and does not affect the actual build of the database. (The main reason we have it, IMHO, is because it is easier to compare changes in this representation than in the more comprehensive XML files that are used to build the tables.)

**Reply by hecon5** [reply]
Correct; I gleaned that after I did my initial export and realizing that "oh hey, there's a LOT of deletions here, what'd I break!?". 

I still think users should be apprised to this...that said...I do think the approach of not exporting linked tables is the correct approach.

**Reply by HughWarrington** [reply]
We could make the old behaviour the default, and just add an option to opt-out of the exports of linked table .SQL files.

**Reply by HughWarrington** [reply]
Having run the new addin (v4.1.0) perhaps I should have been more careful what I wished for... when I export the frontend, it now doesn't output the `tbldefs/*.sql` and `tbldefs/*.json` files for linked tables. This seems like an improvement in the sense that those files are also already exported from my backend, so I am avoiding duplicate files in version control.

Also, on build I no longer see the 'Could not link table' error which is good. However, that's only because the linked table was not created in the front end at all! When my newly built database runs, it immediately crashes because it is trying to read/update a linked table path, i.e. `CurrentDb.TableDefs("MyLinkedTable").Connect`.

> The SQL view is purely for visual reference purposes, and does not affect the actual build of the database.

I have to conclude this appears not to be true? It looks like because the files for linked tables are not exported, those tables are also not generated during build. Perhaps my question should now be, how do you build a database with linked tables in 4.1.0?

I'm sorry for any confusion, I'm still quite new to Access and I might have previously asked for something impossible, namely that a linked table is generated on import which isn't linked to any _particular_ table.

**Reply by hecon5** [reply]
Hmm, something else is going on! I have about a hundred linked tables in one of my applications, and it creates them just fine. 

When you export, can you see if debug VBA errors throws up an error when you export? Wondering if they aren't exported due to an error. 

Can you go to the db with the linked tables (pre rebuild w/o), and do a full export and see if the linked files show up? This should ensure all objects are exported regardless of state.

**Reply by HughWarrington** [reply]
Thinking again, maybe the problem is that `tbldefs/LinkedTable.json` seems to have disappeared from the export along with `tbldefs/LinkedTable.sql` whereas the json file should have been kept?

**Reply by HughWarrington** [reply]
Ah I see what was going on. At the time I ran the export of the frontend, the linked tables existed in the frontend but were not linked to an actual backend file. They were pointing at a non-existent .mdb path. I guess this means they are skipped on export.

I used the frontend to load a new backend (i.e. redirecting the linked tables to an .mdb that does exist) and reran the export and all of the linked table `tbldefs/LinkedTable.json` files were created properly (and not the corresponding .sql files any more). I think this means I can go back to my approach https://github.com/joyfullservice/msaccess-vcs-addin/discussions/540#discussioncomment-10948623 of editing the connect string to a relative path that is guaranteed to exist on import.

**Reply by joyfullservice** [reply]
Yes, that should work. There are a few ways you can accomplish this, but what I might suggest is adding an `AfterExport` hook to programmatically load the `tbldefs/LinkedTable.json` file and check the `items/connect` property to ensure that it is using a relative path to your back-end repo. If it is anything other than `;DATABASE=backend.mdb`, then update it in the json file.

This approach has the advantage that it is super fast, (much faster than relinking the tables in the front end before export) and guarantees that you will always have a buildable front end. You will be able to run the export during development regardless of which back end you are connected to, and it will always generate consistent and correct source files for the linked tables.
